single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -441,9 +443,74 @@ def update_config_and_restart(self) -> None:
             raise ShardingMigrationError(
                 f"Migration of sharding components not permitted, revert config role to {self.state.app_peer_data.role.value}"
             )
-
-        if self.charm.unit.is_leader():
-            self._handle_ldap_config_changes()
+        if not self.charm.unit.is_leader():
+            return
+        self._handle_ldap_config_changes()
+        if system_users_secret_id := self.config.system_users:
+            self.update_internal_users_password_from_secret(system_users_secret_id)
+
+    def update_internal_users_password_from_secret(self, system_users_secret_id: str) -> None:
+        """Validate the secret content and update the password for each user."""
+        if not (
+            self.state.is_role(MongoDBRoles.CONFIG_SERVER)
+            or self.state.is_role(MongoDBRoles.REPLICATION)
+        ):
+            logger.debug("Internal user passwords cannot be managed by shards.")
+            self.state.statuses.add(
+                CharmStatuses.PASSWORD_ON_SHARD.value,
+                scope="app",
+                component=self.name,
+            )
+            return
+        try:
+            user_passwords = self.charm.state.get_secret_from_id(system_users_secret_id)
+            validate_charm_user_password_config(user_passwords)
+        except ModelError as e:
+            logger.error(f"Invalid system-users secret: {e}.")
+            self.charm.status_handler.set_running_status(
+                CharmStatuses.INVALID_SYSTEM_USERS.value,
+                scope="app",
+                statuses_state=self.state.statuses,
+                component_name=self.name,
+            )
+            raise SetPasswordError("The charm has no permission to access the secret.")
+        except (SecretNotFoundError, InvalidPasswordError) as e:
+            logger.error(f"Invalid system-users secret: {e}.")
+            self.charm.status_handler.set_running_status(
```

Comment:
I think this does not need to be a running status, but a regular status: you will exit right after, there's no need to display it immediately, and it's easy to recompute it later.
You'll still be able to remove it as you do below, with the same call.